Added example Tx specification

TimSheard opened the pull request:
# Description

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

TimSheard committed:
Added classes EraSpecPParams, EraSpecTxBody, EraSpecDeleg, EraSpecCert for specifying Era parametric things
Made Era parametric HasSpec instances for all type family instances of Tx TxBody TxWits TxAuxData
Added example Tx specification (bodyspec)

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/ParametricSpec.hs:
```diff
@@ -0,0 +1,301 @@
+  , EraPP era
+  , EraDeleg era
+  ) =>
+  EraOut era fn
```
I think the names of these typeclasses should reflect the `EraTxOut` etc. typeclasses in the implementation. So this typeclass could be named something like `EraSpecTxOut`

TimSheard replied:
That is a good idea, so now we have classes
EraSpecPParams
EraSpecTxOut
EraSpecTxBody
EraSpecDeleg
EraSpecLedger etc.

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/ParametricSpec.hs:
```diff
@@ -0,0 +1,301 @@
+    Term fn (TxOut era) ->
+    Pred fn
+
+betterTxOutShelley ::
```
If this `TxOut` spec is relevant only in the `Shelley` era, then wouldn't it make sense to fix the era to `Shelley` instead of just mentioning the era in the name?

Soupstraw replied:
ok, I see now that you also use the same spec for Allegra

TimSheard replied:
These specs vary by what a Value is, and what a TxOut is. Allegra and Shelly have the same combination.

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/ParametricSpec.hs:
```diff
@@ -0,0 +1,301 @@
+  Term fn (Map (Credential 'Staking (EraCrypto era)) (KeyHash 'StakePool (EraCrypto era))) ->
+  Term fn (ShelleyTxOut era) ->
+  Pred fn
+betterTxOutMary delegs txOut =
```
These specs seem to have a lot of constraints just copy-pasted over, I think these constraints should be factored out into a separate function to be reused across the specs

TimSheard replied:
This is difficult, because two things are varying, Value and TxOut, so it is hard to make a general type that captures all the variations. I will think further on this

TimSheard replied:
I am going to leave this for another PR

Soupstraw commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/ParametricSpec.hs:
```diff
@@ -0,0 +1,301 @@
+-- =========================================================================
+-- Making Cert Era parametric with the EraCert class
+
+class (Era era, HasSpec ConwayFn (TxCert era)) => EraCert era where
```
I think this should also have a better name, `EraSpecTxCert` maybe?

TimSheard replied:
done

TimSheard committed:
Added second TxBody spec example.
All Conway HasSpec instances and EraSpecXX instances are exported
  by Test.Cardano.Ledger.Constrained.Conway.Instances
Added LedgerType.Tests

MaximilianAlgehed commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Cert.hs:
```diff
@@ -40,21 +66,200 @@ certStateSpec =
+        match gdc $ \ [var|gkh|] [var|vkh|] [var|hashVrf|] ->
+          [ dependsOn gdc gkh
+          , dependsOn gdc vkh
+          , dependsOn gdc hashVrf
```
This is unnecessary

MaximilianAlgehed replied:
`match` should introduce these dependencies on its own.

TimSheard replied:
OK, here it is obvious, but another way to think about it, is that it is documentation. Dependency problems are hardest errors to understand. So I would encourage users to use them liberally, if they have some particular order in mind. And not to worry about if they are redundant.

MaximilianAlgehed commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Cert.hs:
```diff
@@ -40,21 +66,200 @@ certStateSpec =
+   in (vrfKeyHashes, coldKeyHashes)
+
+test3 :: forall era. (EraSpecDeleg era, EraSpecPParams era) => IO ()
+test3 = do
```
I'm not a hughe fan of this name for a function... Isn't it better to add an actual test somewhere?

TimSheard replied:
So now it is an actual test. Sometimes I write these just to be sure things work as I develop a specification. Then I forget to turn them into a test. So now it is called.
```
testGenesisCert :: forall era. (AtMostEra BabbageEra era,EraSpecDeleg era, EraSpecPParams era) => Gen Property
```

MaximilianAlgehed commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Cert.hs:
```diff
@@ -40,21 +66,200 @@ certStateSpec =
+          RetirePool {} -> "RetirePool"
+        ShelleyTxCertGenesisDeleg _ -> "Genesis"
+        ShelleyTxCertMir _ -> "Mir"
+  pure (classify True tag (property (conformsToSpec ans spec)))
```
Names...

TimSheard replied:
So I renamed these to testShelleyCert and testConwayCert. I also added the constraint 
```
testShelleyCert :: forall era. (AtMostEra BabbageEra era, EraSpecPParams era, EraSpecDeleg era) => Gen Property
```
Because this test makes no sense in the Conway era.

MaximilianAlgehed commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/TxBodySpec.hs:
```diff
@@ -0,0 +1,322 @@
+        , dependsOn shelleyBody outputs
+        , dependsOn feeInput inputUtxo
+        , dependsOn inputs inputUtxo
+        , dependsOn outputs inputUtxo
```
Many of these are not necessary as they are introduced by match

TimSheard replied:
Please just think of these as documentation.

MaximilianAlgehed commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/TxBodySpec.hs:
```diff
@@ -0,0 +1,322 @@
+  generate $
+    genFromSpec @ConwayFn @(Map (TxIn (EraCrypto era)) (TxOut era)) (hasSize (rangeSize 7 15))
+
+-- | Exercise the 'bodyspec'
+go ::
+  forall era.
+  ( EraSpecTxOut era ConwayFn
+  , EraSpecCert era ConwayFn
+  , HasSpec ConwayFn (Tx era)
+  ) =>
+  IO ()
+go = do
+  utxo <- UTxO <$> testUTxO @era
+  certState <-
+    generate $
+      genFromSpec @ConwayFn @(CertState era)
+        (certStateSpec @ConwayFn @era) -- (lit (AccountState (Coin 1000) (Coin 100))) (lit (EpochNo 100)))
+        -- error "STOP"
+  certsEnv <- generate $ genFromSpec @ConwayFn @(CertsEnv era) certsEnvSpec
+
+  (basic, subutxo, feeinput) <-
+    generate $ genFromSpec (bodyspec @era @ConwayFn utxo certsEnv certState)
+
+  putStrLn
+    ("Input UTxO, total " ++ show (coinBalance @era utxo) ++ ", size = " ++ show (Map.size (unUTxO utxo)))
+  putPretty "\nSubMap of Utxo" (UTxO subutxo)
+  putStrLn ("SubMap of UTxO, total Coin " ++ show (coinBalance @era (UTxO subutxo)))
+  putPretty "\nfeeInput" feeinput
+  putPretty "\nTxBody" basic
```
Another bad name for a test that perhaps should just be an actual test.

TimSheard replied:
So all these things have been made into real tests, or removed.

MaximilianAlgehed commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/TxBodySpec.hs:
```diff
@@ -0,0 +1,322 @@
+        , assert $ sizeOf_ outputs ==. 4
+        ]
+
+baz :: EraTxOut era => TxOut era -> TxOut era
+baz x = x & coinTxOutL .~ ((x ^. coinTxOutL) <+> (Coin 100))
+
+go2 ::
+  forall era.
+  ( EraSpecTxOut era ConwayFn
+  , EraSpecCert era ConwayFn
+  , HasSpec ConwayFn (Tx era)
+  ) =>
+  IO ()
```
Same issue as above with the names of tests

TimSheard replied:
so added to LedgerTypes.Tests with a better name. An all those tests are run Nightly
